src/ddspdrum/module.py

```diff
@@ -302,19 +386,18 @@ def __init__(
         phase: float = 0,
     ):
         super().__init__(midi_f0=midi_f0, mod_depth=mod_depth, phase=phase)
-
-        # 0 is square. 1 is saw.
-        assert 0.0 <= shape <= 1.0
-        self.shape = shape
+        self.add_parameter("shape", shape, 0, 1)
 
     def oscillator(self, argument):
         square = np.tanh(np.pi * self.k * np.sin(argument) / 2)
-        return (1 - self.shape / 2) * square * (1 + self.shape * np.cos(argument))
+        shape = self.parameters['shape'].value
+        return (1 - shape / 2) * square * (1 + shape * np.cos(argument))
 
     @property
     def k(self):
         # What does k mean here? Can we give it a better name?
-        f0 = midi_to_hz(self.midi_f0 + self.mod_depth)
+        pitch = self.parameters['pitch'].value + self.parameters['mod_depth'].value
```

Comment:
"Pitch" might be a little misleading here.

`k` is used to determine how many partials to calculate, and therefore is dependant on f0 and Nyquist. Higher f0, less partials.

`f0` is meant to calculate number-of-partials for a worst-case scenario. I.e., the pitch at the highest point in its modulation.

Reply:
Yes makes sense -- will rename to something more meaningful.